controller_interface/test/controller_base_test.cpp

```diff
@@ -0,0 +1,313 @@
+  EXPECT_CALL(controller, initRequest(_, _, _, _))
+      .Times(1)
+      .WillOnce(DoAll(InvokeWithoutArgs(&controller, &ControllerMock::initializeState),
+                      Return(true)));
+
+  // not initialized
+  ASSERT_FALSE(controller.stopRequest(time));
+
+  ASSERT_TRUE(controller.initRequest(robot_hw, *node_handle, *node_handle, resources));
+  ASSERT_TRUE(controller.stopRequest(time));
+
+  // can stop multiple times
+  ASSERT_TRUE(controller.stopRequest(time));
+}
+
+TEST(ControllerBaseAPI, WaitRequestTest)
+{
+  StrictMock<ControllerMock> controller;
+
+  hardware_interface::RobotHW* robot_hw = nullptr;
+  ros::NodeHandle* node_handle = nullptr;
+  ControllerMock::ClaimedResources resources;
+  const ros::Time time;
+
+  EXPECT_CALL(controller, waiting(_)).Times(2);
+  EXPECT_CALL(controller, initRequest(_, _, _, _))
+      .Times(1)
+      .WillOnce(DoAll(InvokeWithoutArgs(&controller, &ControllerMock::initializeState),
+                      Return(true)));
+
+  // not initialized
+  ASSERT_FALSE(controller.waitRequest(time));
+
+  ASSERT_TRUE(controller.initRequest(robot_hw, *node_handle, *node_handle, resources));
+  ASSERT_TRUE(controller.waitRequest(time));
+
+  // can wait multiple times
+  ASSERT_TRUE(controller.waitRequest(time));
+}
+
+TEST(ControllerBaseAPI, AbortRequestTest)
+{
+  StrictMock<ControllerMock> controller;
+
+  hardware_interface::RobotHW* robot_hw = nullptr;
+  ros::NodeHandle* node_handle = nullptr;
+  ControllerMock::ClaimedResources resources;
+  const ros::Time time;
+
+  EXPECT_CALL(controller, aborting(_)).Times(2);
+  EXPECT_CALL(controller, initRequest(_, _, _, _))
+      .Times(1)
+      .WillOnce(DoAll(InvokeWithoutArgs(&controller, &ControllerMock::initializeState),
+                      Return(true)));
+
+  // not initialized
+  ASSERT_FALSE(controller.abortRequest(time));
+
+  ASSERT_TRUE(controller.initRequest(robot_hw, *node_handle, *node_handle, resources));
+  ASSERT_TRUE(controller.abortRequest(time));
+
+  // can abort multiple times
+  ASSERT_TRUE(controller.abortRequest(time));
+}
+
+TEST(ControllerBaseAPI, UpdateRequestTest)
+{
+  StrictMock<ControllerMock> controller;
+
+  hardware_interface::RobotHW* robot_hw = nullptr;
+  ros::NodeHandle* node_handle = nullptr;
+  ControllerMock::ClaimedResources resources;
+  const ros::Time time;
+  const ros::Duration period;
+
+  EXPECT_CALL(controller, starting(_)).Times(1);
+  EXPECT_CALL(controller, update(_, _)).Times(1);
+  EXPECT_CALL(controller, initRequest(_, _, _, _))
+      .Times(1)
+      .WillOnce(DoAll(InvokeWithoutArgs(&controller, &ControllerMock::initializeState),
+                      Return(true)));
+
+  // update is not called unless state is RUNNING
+  controller.updateRequest(time, period);
+
+  ASSERT_TRUE(controller.initRequest(robot_hw, *node_handle, *node_handle, resources));
+  ASSERT_TRUE(controller.startRequest(time));
+
+  controller.updateRequest(time, period);
+}
+
+int main(int argc, char** argv)
```

Comment:
If you aren't planning on extending this, you may link to `${GMOCK_MAIN_LIBRARIES}` and drop the definition of the main function here.

Reply:
Done.